Add a Docker based development environment

gjnoonan opened the pull request:
Whilst this works, and is what I have been running locally, it is not *quite* ready to go into master yet; I still need to rebase and split out the commits. However I thought I would push this up as a draft to allow people to raise comment/suggestions.

I've used alpine as the container as it is lightweight and produces a nice small image and is generally really good for this application. The main error that comes up at the moment is:

```
Notice: iconv(): Wrong charset, conversion from `ISO-8859-1' to `ascii//TRANSLIT' is not allowed in /var/www/functions.php on line 1033

Warning: Cannot modify header information - headers already sent by (output started at /var/www/functions.php:1033) in /var/www/index.php on line 2884
```
I know why the first one errs and the second one can be ignored as it only displays because of the first. At the moment it doesn't stop anything from working. I may switch alpine out for debian if it proves a pain

I've added a Makefile for convenience, I realise that a Powershell file may need to be included too (or indeed a VSCode .devcontainer) for those running windows, but I've had no chance to write/test one. 

When you first clone the  project and run `make install` it will 
1) copy the env.example file to .env
2)Build the docker containers. unless you have deleted the image, it has changed,  or you have run `reset` this will only happen the first time and will be quite slow.
3) start all the containers in the background.

From here all you need to do is visit the webpage via port 8080, upload your Collection.XML via the usual method and then everything is ready to go. You can edit your PHP files locally and the changes will be reflected.

during development you _should_ not have to run `make install` again, just use `make start` and `make stop`

MySQL will automatically create the database, and run schema.sql to set-up the required structure once the container has been started.

If you don't want to, or can't run the Makefile then you can run the steps manually:

1) docker pull --parallel 
2) docker build --no-cache --pull  (the pull here makes sure you have the latest base image and won't use the local one if it exists)
3) docker-compose up
4) profit

On a side note, as part of another patch series/MR I would like introduce a standard Collection.xml fixture into the database to use for development. I have one that I used locally when developing #2. This contains a few Blu-Rays and UltraHD titles; However, if we add a DVD, HD-DVD, as well as a release or two in non-english languages we should have enough to make sure nothing is breaking :-)

**note: If you would like to test this out, in your localsiteconfig.php file set `$dbhost = 'db';`**

_this wall of text will be cleaned up into something more cohesive before it is taken out of draft_

julmud commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+        libpng-dev \
+        jpeg-dev \
+        libjpeg \
+        libjpeg-turbo-dev \
```
Wouldn't this solve the iconv problem before going the full Debian way?
```suggestion
        libjpeg-turbo-dev \
        gnu-libiconv \
```

gjnoonan replied:
Yep, that was my "I know why this one errs", mentioning debian was about something different, In my defence, I did say "this wall of text will be cleaned up into something more cohesive before it is taken out of draft" 🤣  I Should have had another coffee before pushing that up.

Have you tested this MR, does the workflow Work fine for you?

julmud replied:
I've now installed a Docker environment on my Windows laptop and have been able to test it. I had to tweak a few things to make it work, but other than that, it looks really good.

Here's what I had to do to make it run under Windows (without make...):

1. Install Docker Desktop
2. Activate the Shared Drives in the settings of Docker Desktop
3. Create a `.env` file based on the `.env.example` file (BTW, the `.env` file should be added to `.gitignore`)
4. Don't run `docker pull --parallel`, as it doesn't work
5. Run `docker build --rm --no-cache --pull -f Dockerfile -t phpdvdprofiler:latest .`
6. Run `docker-compose up`
7. Profit!

Thanks a lot!

julmud commented:
No, I couldn't test it, as I currently don't have a running Docker
environment anywhere.... :-( But on paper it looks good. ;-)

julmud commented:
I've made a configuration which adds:

- xdebug to the app container
- a fix for the missing iconv library
- a VSCode .devcontainer which enables debugging directly from VSCode

Shall I simply push my commits to this branch and merge request?

(Alas, I'm having some trouble with line endings to be able to easily run git from the containers and VSCode...)

gjnoonan commented:
Sorry @julmud, it has been a busy week, I've not found as much "free" time as I'd hoped.

My Local branch also fixes the iconv issue. Good to hear you did the .devcontainer, I also started that but did not get it to a point where I was happy with it.

Yep please either push to the branch or send me the patches. I will rebase the commits and provide correct attribution before we merge.

julmud commented:
@gjnoonan I've been unsuccessful in pushing directly the commits to your branch or this pull request, so I've sent you an email with the patches. ;-)

gjnoonan committed:
dev: add docker based development environment

To aid with development speed, as well as to ensure consistency, a
standardised development environment has been created.

We also provide a Makefile to take some more pain away which provides
the following targeets:

  * `make install` - create .env, download dependencies, create the db, start
	everything up
  * `make start`  - start the environment
  * `make stop`   - tear it all down
  * `make reset` kill all the containers and run `install`

Signed-off-by: Gavin-John Noonan <mail@gjnoonan.co.uk>

gjnoonan commented:
@julmud I reviewed your patches earlier, via email :-) if you could rebase against my latest push then they should apply cleanly.

julmud commented:
> @julmud I reviewed your patches earlier, via email :-) if you could rebase against my latest push then they should apply cleanly.

@gjnoonan Is it normal that the `docker-compose.yml` file is missing from your latest commits? And the `.env.example` file seems to be missing too.

gjnoonan commented:
> > @julmud I reviewed your patches earlier, via email :-) if you could rebase against my latest push then they should apply cleanly.
> 
> @gjnoonan Is it normal that the `docker-compose.yml` file is missing from your latest commits? And the `.env.example` file seems to be missing too.

hmm sorry about that, I must not have paid enough attention after my last rebase.

julmud committed:
Fix the missing dependency for iconv

With the default alpine image, the iconv library is wrongly configured,
resulting in a PHP notice (iconv(): Wrong charset, conversion from
`ISO-8859-1' to `ascii//TRANSLIT' is not allowed in
/var/www/functions.php on line 1033) when running an update. Adding the
GNU libiconv library to the image solves this problem.

julmud committed:
Add container configuration for Visual Studio Code

The devcontainer allows us to run everything directly in containers
from the Visual Studio Code GUI. Adding git to the base image, so that
all version control operations can be done directly when running the
containers from Visual Studio Code.

julmud committed:
Add xdebug and a Visual Studio Code launch config

For easier debugging with Visual Studio Code, add the
xdebug extension to the PHP installation, and a configuration
for debugging the project from the GUI. Note that xdebug
connects to port 9229, as the default port of 9000 is
already being used by php-fpm.
